dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java

```diff
@@ -0,0 +1,174 @@
+        } else {
+            log.error("workerGroupTaskDispatcher {} not found, will set task {} fail",
+                    workerGroup, taskExecutionRunnable.getTaskInstance().getId());
+        }
+        return false;
+    }
+
+    /**
+     * Stops a specific worker group's task dispatch waiting queue looper.
+     *
+     * @param workerGroup the identifier for the worker group
+     */
+    private synchronized void deleteWorkerGroup(String workerGroup) {
+        WorkerGroupTaskDispatcher workerGroupTaskDispatcher = dispatchWorkerMap.get(workerGroup);
+        if (workerGroupTaskDispatcher != null) {
+            workerGroupTaskDispatcher.markDispatcherClosing();
+        } else {
+            log.warn("workerGroupTaskDispatcher {} not found", workerGroup);
+        }
+    }
+
+    /**
+     * add workerGroup
+     *
+     * @param workerGroup the identifier for the worker group
+     */
+    private synchronized void addWorkerGroup(String workerGroup) {
+        WorkerGroupTaskDispatcher workerGroupTaskDispatcher = dispatchWorkerMap.get(workerGroup);
+        if (workerGroupTaskDispatcher == null) {
+            workerGroupTaskDispatcher = new WorkerGroupTaskDispatcher(workerGroup, taskExecutorClient);
+            dispatchWorkerMap.put(workerGroup, workerGroupTaskDispatcher);
+            workerGroupTaskDispatcher.start();
+        } else {
+            workerGroupTaskDispatcher.markDispatcherStart();
+        }
+    }
+
+    /**
+     * Stop all workerGroupTaskDispatchWaitingQueueLooper
+     */
+    @Override
+    public void close() throws Exception {
+        log.info("WorkerGroupTaskDispatcherManager start close");
+        for (Map.Entry<String, WorkerGroupTaskDispatcher> entry : dispatchWorkerMap.entrySet()) {
+            try {
+                entry.getValue().markDispatcherClosing();
+            } catch (Exception e) {
+                log.error("stop worker group error", e);
+            }
+        }
+        log.info("WorkerGroupTaskDispatcherManager closed");
+    }
+
+    @Override
+    public void onWorkerGroupAdd(List<WorkerGroup> workerGroups) {
+        this.checkAndRemoveClosedDispatchWorker();
+        for (WorkerGroup workerGroup : workerGroups) {
+            this.addWorkerGroup(workerGroup.getName());
+        }
+    }
+
+    @Override
+    public void onWorkerGroupChange(List<WorkerGroup> workerGroups) {
+        this.checkAndRemoveClosedDispatchWorker();
+        String workerGroupsString = workerGroups.stream()
+                .map(WorkerGroup::getName)
+                .collect(Collectors.joining(", "));
+        log.info("Worker groups: {}", workerGroupsString);
+    }
+
+    @Override
+    public void onWorkerGroupDelete(List<WorkerGroup> workerGroups) {
+        for (WorkerGroup workerGroup : workerGroups) {
+            try {
+                this.deleteWorkerGroup(workerGroup.getName());
+            } catch (Exception e) {
+                log.error("stop worker group error", e);
```

Comment:
```suggestion
                log.error("Delete worker group: {} from WorkerGroupTaskDispatcherManager error", workerGroup.getName(), e);
```

Reply:
Thank you very much for your suggestions! I completely agree with your changes and they do make the code clearer. I will make the changes according to your suggestions.